classes/observers/listeners/UsageEventLog.inc.php

```diff
@@ -0,0 +1,472 @@
+        $context = $usageEvent->context;
+
+        // Geo data
+        [$country, $region, $city] = $this->getGeoData($site, $context, $ip, $hashedIp, $flushCache);
+
+        // institutions IDs
+        $institutionIds = [];
+        if ($context->isInstitutionStatsEnabled($site)) {
+            $institutionIds = $this->getInstitutionIds($context->getId(), $ip, $hashedIp, $flushCache);
+        }
+
+        // format the usage event log entry
+        $usageEventLogEntry = [
+            'time' => $usageEvent->time,
+            'ip' => $hashedIp,
+            'userAgent' => $request->getUserAgent(),
+            'canonicalUrl' => $usageEvent->canonicalUrl,
+            'assocType' => $usageEvent->assocType,
+            'contextId' => $context->getId(),
+            'submissionId' => $usageEvent->submission?->getId(),
+            'representationId' => $usageEvent->representation?->getId(),
+            'submissionFileId' => $usageEvent->submissionFile?->getId(),
+            'fileType' => $usageEvent->submissionFile ? $this->getDocumentType($usageEvent->submissionFile->getData('mimetype')) : null,
+            'country' => $country,
+            'region' => $region,
+            'city' => $city,
+            'institutionIds' => $institutionIds,
+            'version' => $usageEvent->version
+        ];
+        // get application specific IDs
+        if (Application::get()->getName() == 'ojs2') {
+            $usageEventLogEntry['issueId'] = $usageEvent->issue?->getId();
+            $usageEventLogEntry['issueGalleyId'] = $usageEvent->issueGalley?->getId();
+        } elseif (Application::get()->getName() == 'omp') {
+            $usageEventLogEntry['chapterId'] = $usageEvent->chapter?->getId();
+            $usageEventLogEntry['seriesId'] = $usageEvent->series?->getId();
+        }
+        return $usageEventLogEntry;
+    }
+
+    /**
+     * Get the path to the Geo DB file.
+     */
+    protected function getGeoDBPath(): string
+    {
+        return StatisticsHelper::getUsageStatsDirPath() . '/IPGeoDB.mmdb';
+    }
+
+    /**
+     * Get the path to the salt file.
+     */
+    protected function getSaltFileName(): string
+    {
+        return StatisticsHelper::getUsageStatsDirPath() . '/salt';
+    }
+
+    /**
+     * Get current day usage event log file name.
+     */
+    protected function getUsageEventLogFileName(): string
+    {
+        $usageEventLogsDir = StatisticsHelper::getUsageStatsDirPath() . '/usageEventLogs';
+        if (!file_exists($usageEventLogsDir) || !is_dir($usageEventLogsDir)) {
+            $fileMgr = new PrivateFileManager();
+            $success = $fileMgr->mkdirtree($usageEventLogsDir);
+            if (!$success) {
+                // Files directory wrong configuration?
+                error_log("UsageEventLog: Couldn't create {$usageEventLogsDir}.");
+            }
+        }
+        return $usageEventLogsDir . '/usage_events_' . date('Ymd') . '.log';
+    }
+
+    /**
+     * Get salt
```

Comment:
Could you document this a little more? If I'm not sure what salt is or how it's used in event logging, this is confusing.

Reply:
Does this sound OK:
  ```
/**
  * Get salt (used for IP hashing) from the file.
  * For privacy reasons, the salt will change every day.
  * When salt changes, the IP hash will change too, so the cach containing the hashed IPs (for Geo and institutions data) will be removed.
  * If the salt file does not exist, it will be created.
  * If the salt file exists and it was not modified at the current day the salt will be changed.
*/
```